### PR TITLE
Add ability for lists to be unordered, plus a few other fixes

### PR DESCRIPTION
**Pull request by timwoj**
@JustinAzoff and I noticed the other day that removing an element from a list can be an O(n) operation, since it will remove an element and then move all of the items in the list down to fill the new hole in the list. This doesn't necessarily need to happen if the creator of the list doesn't care what order the elements are in after removal. This case is true of the list of timers that `Conn` holds, for example.

This PR adds the ability for lists to be unordered at definition-time, along with a bunch of unit testing that was missing for `List`. It also fixes a bug with `List::replace()` if you tried to use it with a type that wasn't a pointer (i.e. not a `PList`).

### CODE REVIEW

**Comment by JustinAzoff** [comment]
where you also going to use this to change the connection timer list to be unordered?

**Comment by timwoj** [comment]
`Connection` uses `timer_list`, so changing that type definition in `BroList` covered it.

**Comment by JustinAzoff** [comment]
ah, I see.  I  think with my fix this wouldn't end up being needed since the timer list shouldn't grow larger than 3, but it'll probably come in handy for other things.

File: src/List.h [hunk 1]
```diff
@@ -29,7 +29,9 @@
 // TODO: this can be removed in v3.1 when List::sort() is removed
 typedef int (*list_cmp_func)(const void* v1, const void* v2);
 
-template<typename T>
+enum list_order { LIST_ORDERED, LIST_UNORDERED };
```
Comment (jsiwek) [inline src/List.h:32]:
```suggestion
enum ListOrder { LIST_ORDERED, LIST_UNORDERED };
```

Think that's the naming convention.

`enum class` may also be preferable, in general.

Comment (timwoj) [reply src/List.h:32]: I was following the naming scheme used by `dict_order` there, but that could probably be renamed too.

File: src/List.h [hunk 2]
```diff
@@ -287,6 +287,20 @@ class List {
...
+		if constexpr ( std::is_pointer_v<T> )
+			return nullptr;
+		else if constexpr ( std::is_integral_v<T> )
+			return 0;
+		else if constexpr ( std::is_floating_point_v<T> )
+			return 0.0;
+		else
+			// TODO: Should this return an error at compile time that we don't
+			// know what this type is?
+			return T();
```
[the remark this answers is not in the document]
Comment (timwoj) [reply src/List.h:313]: I think that's true for scalar types and pointers but not for classes. Zero-initialization using `T{}` for a class zero-initializes all of the internal values for the class but doesn't actually call any of the constructors, at least according to https://en.cppreference.com/w/cpp/language/zero_initialization

[the remark this answers is not in the document]
Comment (jsiwek) [reply src/List.h:313]: `T{}` or `T()` would be starting with [value initialization](https://en.cppreference.com/w/cpp/language/value_initialization) rules, no?  That may end up doing just zero-initialization in some cases, but others do end up calling a ctor via a default-initialization step.

[the remark this answers is not in the document]
Comment (timwoj) [reply src/List.h]: Yep, read through it again and calling it with {} should be fine for everything.

**Pull request merged**